textattack/search_methods/genetic_algorithm.py

```diff
@@ -61,19 +61,18 @@ def _perturb(self, pop_member, original_result, index=None):
         Returns:
             Perturbed `PopulationMember`
         """
-        num_words = pop_member.num_replacements_per_word.shape[0]
-        num_replacements_per_word = np.copy(pop_member.num_replacements_per_word)
-        non_zero_indices = np.count_nonzero(num_replacements_per_word)
+        num_words = pop_member.attacked_text.num_words
+        # `word_attr_list` contains some attribute of each word that helps us choose which word to transform.
+        word_attr_list = np.copy(getattr(pop_member, self._attr_name))
+        non_zero_indices = np.count_nonzero(word_attr_list)
         if non_zero_indices == 0:
             return pop_member
         iterations = 0
         while iterations < non_zero_indices:
             if index:
                 idx = index
             else:
-                w_select_probs = num_replacements_per_word / np.sum(
-                    num_replacements_per_word
-                )
+                w_select_probs = word_attr_list / np.sum(word_attr_list)
```

Comment:
Instead of doing the `word_attr_list` things depending on `self.atr_name`, I think we should:
- Get rid of the `self.attr_name` thing, it feels like not great practice and hard to read/understand.
- Make getting the selection probs into a separate function so it can be overridden by the `Alzantot` and `Improved` versions.
- Make all `PopulationMember`s have both of the attribute lists, but which attribute gets used depends on the specifics of select probs as described above.

Reply:
I really like those changes, but one problem is that `word_attr_list` is actually a copy of the attribute list and used as temporary variable within the while loop.  So it's unclear how this works with the word selection prob function. We can't create copy of the attribute list within the function and use it since its scope is not just limited to the function.